graphql-cli/src/main/java/io/ballerina/graphql/cmd/GraphqlCmd.java

```diff
@@ -168,57 +198,116 @@ private void validateInputFlags() throws CmdException {
+    }
+
+    private boolean isModeCompatible() throws CmdException {
+        String filePath = argList.get(0);
+        if (mode != null) {
+            if (MODE_CLIENT.equals(mode)) {
+                return filePath.endsWith(Constants.YAML_EXTENSION) || filePath.endsWith(Constants.YML_EXTENSION);
+            } else if (MODE_SCHEMA.equals(mode)) {
+                return filePath.endsWith(Constants.BAL_EXTENSION);
+            } else if (MODE_SERVICE.equals(mode)) {
+                return filePath.endsWith(Constants.GRAPHQL_EXTENSION);
+            } else {
+                throw new CmdException(String.format(MESSAGE_FOR_INVALID_MODE, mode));
+            }
+        } else if (filePath.endsWith(Constants.GRAPHQL_EXTENSION)) {
+            throw new CmdException(String.format(MESSAGE_FOR_GRAPHQL_FILE_WITH_NO_MODE, filePath));
+        }
+        return true;
     }
 
     /**
      * Execute the correct operation according to the given inputs.
      *
-     * @throws CmdException               when a graphql command related error occurs
-     * @throws ParseException             when a parsing related error occurs
-     * @throws IOException                If an I/O error occurs
-     * @throws GenerationException        when a graphql client generation related error occurs
-     * @throws ValidationException        when validation related error occurs
-     * @throws SchemaFileGenerationException  when a SDL schema generation related error occurs
+     * @throws CmdException                  when a graphql command related error occurs
+     * @throws ParseException                when a parsing related error occurs
+     * @throws IOException                   If an I/O error occurs
+     * @throws GenerationException           when a graphql client generation related error occurs
+     * @throws ValidationException           when validation related error occurs
+     * @throws SchemaFileGenerationException when a SDL schema generation related error occurs
      */
     private void executeOperation()
             throws CmdException, ParseException, IOException, ValidationException, GenerationException,
             SchemaFileGenerationException {
         String filePath = argList.get(0);
-        if (filePath.endsWith(YAML_EXTENSION) || filePath.endsWith(YML_EXTENSION)) {
+
+        if ((MODE_CLIENT.equals(mode) || mode == null) &&
+                (filePath.endsWith(YAML_EXTENSION) || filePath.endsWith(YML_EXTENSION))) {
+            setClientCodeGenerator(new ClientCodeGenerator());
             generateClient(filePath);
-        } else if (filePath.endsWith(BAL_EXTENSION)) {
+        } else if ((MODE_SCHEMA.equals(mode) || mode == null) && (filePath.endsWith(BAL_EXTENSION))) {
             generateSchema(filePath);
-        } else {
-            throw new CmdException(MESSAGE_FOR_INVALID_FILE_EXTENSION);
+        } else if ((MODE_SERVICE.equals(mode)) && (filePath.endsWith(GRAPHQL_EXTENSION))) {
```

Comment:
Can't we have the same logic in the `if` block, as client and schema generation?
```suggestion
        } else if ((MODE_SERVICE.equals(mode) || mode == null) && (filePath.endsWith(GRAPHQL_EXTENSION))) {
```

Reply:
There can be new modes introduced for a file with `.graphql` extension. So it was decided to make the mode flag compulsory for a file with `.graphql` extension.